src/internal/arg.c

```diff
@@ -84,3 +84,98 @@ sexp* rlang_is_missing(sexp* _call, sexp* _op, sexp* args, sexp* env) {
 
   return r_lgl(r_eval(r_x_sym, env) == r_missing_sym);
 }
+
+static sexp* stop_arg_match_call = NULL;
+void arg_match0_abort(const char* msg, sexp* arg_nm_promise, sexp* env);
+sexp* rlang_ext2_arg_match0(sexp* _call, sexp* _op, sexp* args, sexp* env) {
+  args = r_node_cdr(args);
+
+  sexp* arg =            KEEP(r_eval(r_node_car(args), env)); args = r_node_cdr(args);
+  sexp* values =         KEEP(r_eval(r_node_car(args), env)); args = r_node_cdr(args);
+  // Expensive by default, forced only if necessary.
+  sexp* arg_nm_promise = r_node_car(args);
+
+  if (r_typeof(arg) != r_type_character) {
+    arg_match0_abort("`%s` must be a character vector.", arg_nm_promise, env);
+  }
+  if (r_typeof(values) != r_type_character) {
+    r_abort("`values` must be a character vector.");
+  }
+
+  r_ssize len = r_length(arg);
+  r_ssize len_values = r_length(values);
+  if (len != 1 && len != len_values) {
+    //arg_match0_abort("`%s` must be a string or have the same length as `values`.",  arg_nm_promise, env);
+
+    arg = Rf_ScalarString(STRING_ELT(arg, 0));
+    r_eval_with_xyz(stop_arg_match_call, r_base_env, arg, values, arg_nm_promise);
+
+    while (1); // No return
+  }
+
+  if (len == 1) {
+    sexp* arg_char = STRING_ELT(arg, 0);
+    for (r_ssize i = 0; i < len_values; ++i) {
+      if (arg_char == STRING_ELT(values, i)) {
+        FREE(2);
+        return(Rf_ScalarString(arg_char));
+      }
+    }
+
+    r_eval_with_xyz(stop_arg_match_call, r_base_env, arg, values, arg_nm_promise);
+
+    while (1); // No return
+  } else {
+    bool need_match = false;
+    r_ssize ii = 0;
+    for (; ii < len; ++ii) {
+      if (STRING_ELT(arg, ii) != STRING_ELT(values, ii)) {
+        need_match = true;
+        break;
+      }
+    }
+
+    // Elements are in order, return first
+    if (!need_match) {
+      FREE(2);
+      return(Rf_ScalarString(STRING_ELT(values, 0)));
+    }
+
+    for (R_xlen_t i = ii; i < len; ++i) {
+      bool matched = false;
+      for (r_ssize j = ii; j < len; ++j) {
+        if (STRING_ELT(arg, i) == STRING_ELT(values, j)) {
```

Comment:
Is this correct? I think we need to use something equivalent to `all(arg[-1] %in% values)`? This can probably be moved to R code if needed, just add a `sexp* call = r_parse(*code*)` in the init section and use `r_eval_with_xy()`. Alternatively, I think we can call `match()` in a loop with the C API.

Reply:
I think we need `identical(sort(arg), sort(values))`. Will revisit.

Reply:
Good point!